pkg/scheduler/queue/queue.go

```diff
@@ -128,50 +128,35 @@ func (q *RequestQueue) GetNextRequestForQuerier(ctx context.Context, last UserIn
 	q.mtx.Lock()
 	defer q.mtx.Unlock()
 
-	querierWait := false
-
-FindQueue:
-	// We need to wait if there are no users, or no pending requests for given querier.
-	for (q.queues.len() == 0 || querierWait) && ctx.Err() == nil && !q.stopped {
-		querierWait = false
-		q.cond.Wait(ctx)
-	}
-
-	if q.stopped {
-		return nil, last, ErrStopped
-	}
-
-	if err := ctx.Err(); err != nil {
-		return nil, last, err
-	}
-
 	for {
+		if q.stopped {
+			return nil, last, ErrStopped
+		}
+		if err := ctx.Err(); err != nil {
+			return nil, last, err
+		}
+
 		queue, userID, idx := q.queues.getNextQueueForQuerier(last.last, querierID)
 		last.last = idx
 		if queue == nil {
-			break
+			// no queues for this querier
+			continue
```

Comment:
it seems like we should `return nil, last, err` here; we're holding the mutex lock which guards all changes to the queues, and there's no queues for this querier, wouldn't everyone be happier if we returned instead of waiting for something that will never come?

the loop will eventually just hit the the `ctx.Err()` case and  `return nil, last, err` anyway, but after waiting longer and doing a bunch more checks.